src/target/riscv/riscv.c

```diff
@@ -3987,36 +3987,42 @@ int riscv_enumerate_triggers(struct target *target)
 		tselect_rb &= ~(1ULL << (riscv_xlen(target) - 1));
 		if (tselect_rb != t)
 			break;
-		uint64_t tdata1;
-		result = riscv_get_register(target, &tdata1, GDB_REGNO_TDATA1);
-		if (result != ERROR_OK)
-			return result;
-
-		int type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
-		if (type == 0)
-			break;
-		switch (type) {
-			case 1:
-				/* On these older cores we don't support software using
-					* triggers. */
-				riscv_set_register(target, GDB_REGNO_TDATA1, 0);
-				break;
-			case 2:
-				if (tdata1 & CSR_MCONTROL_DMODE(riscv_xlen(target)))
-					riscv_set_register(target, GDB_REGNO_TDATA1, 0);
-				break;
-			case 6:
-				if (tdata1 & CSR_MCONTROL6_DMODE(riscv_xlen(target)))
-					riscv_set_register(target, GDB_REGNO_TDATA1, 0);
-				break;
-		}
 
 		uint64_t tinfo;
 		result = riscv_get_register(target, &tinfo, GDB_REGNO_TINFO);
-		if (result == ERROR_OK)
+		if (result == ERROR_OK) {
+			/* tinfo == 0 invalid tinfo
+			 * tinfo == 1 trigger doesn’t exist */
+			if (tinfo == 0 || tinfo == 1)
+				break;
 			r->trigger_tinfo[t] = tinfo;
-		else
+		} else {
+			uint64_t tdata1;
+			result = riscv_get_register(target, &tdata1, GDB_REGNO_TDATA1);
+			if (result != ERROR_OK)
+				return result;
+
+			int type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
+			if (type == 0)
+				break;
+			switch (type) {
+				case 1:
+					/* On these older cores we don't support software using
+						* triggers. */
+					riscv_set_register(target, GDB_REGNO_TDATA1, 0);
+					break;
+				case 2:
+					if (tdata1 & CSR_MCONTROL_DMODE(riscv_xlen(target)))
+						riscv_set_register(target, GDB_REGNO_TDATA1, 0);
+					break;
+				case 6:
+					if (tdata1 & CSR_MCONTROL6_DMODE(riscv_xlen(target)))
+						riscv_set_register(target, GDB_REGNO_TDATA1, 0);
+					break;
+			}
 			r->trigger_tinfo[t] = 1 << type;
+		}
+		LOG_TARGET_DEBUG(target, "Trigger %u: supported types (mask) = 0x%08x", t, r->trigger_tinfo[t]);
```

Comment:
Thank you for the changes. Overall, the code looks good.

One last point: 
Shouldn't we unconditionally disable any trigger found (by writing `0` to `tdata1`)? If so, the code would look as shown below. @timsifive, what is your opinion, please?

```suggestion
		} else {
			uint64_t tdata1;
			result = riscv_get_register(target, &tdata1, GDB_REGNO_TDATA1);
			if (result != ERROR_OK)
				return result;

			int type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
			if (type == 0)
				break;

			r->trigger_tinfo[t] = 1 << type;
		}
		
		/* Successfully enumerated one trigger */
		LOG_TARGET_DEBUG(target, "Trigger %u: supported types (mask) = 0x%08x", t, r->trigger_tinfo[t]);
		
		/* Disable the found trigger. Writing 0 to tdata1 guarantees the trigger will get disabled. */
		riscv_set_register(target, GDB_REGNO_TDATA1, 0);
```

Reply:
> Shouldn't we unconditionally disable any trigger found (by writing 0 to tdata1)? If so, the code would look as shown below. @timsifive, what is your opinion, please?

I don't know. OpenOCD's capability of specifying triggers is fairly limited, so I can see how users would want to set triggers directly using other tools. We should be wary of stomping on them. On the other hand, what happens if OpenOCD crashes and we connect again to an important debug session. Now any triggers that OpenOCD had set will persist and the user won't have a good UI to remove them.

I'd say print a message using LOG_TARGET_INFO when OpenOCD finds a trigger that is set (and owned by the debugger). Then we could add (in a different PR) a command to easily disable triggers that OpenOCD didn't set.

Reply:
@timsifive - Thank you for replying.

> I'd say print a message using LOG_TARGET_INFO when OpenOCD finds a trigger that is set (and owned by the debugger).

This sounds fine. However, detecting an active trigger (for all supported trigger types) sounds like enough work for a separate merge request. For that reason, it seems OK to me to accept this merge request as is. And further improve the logic around triggers in subsequent merge requests, as needed.